### PR TITLE
chore(build): GitHub Action for rebuilding native libs

### PR DESCRIPTION
**Pull request by jerrinot**
GitHub workflow to rebuild  all C++ native libraries and push it to the current branch. 

The idea is to start the workflow manually from the Actions tab, against a branch with a PR. The workflow will then push the newly updated native library to the PR. Eventually, it could be automated further, for example, it could run automatically on all PRs whenever there is a change in `core/src/c`

Assumptions:
1. Github runner has permission to write. If this is not the case, we can create a token specifically for the PUSH action or the workflow could be configured to open a PR instead of a direct commit. but that feels like overkill to me. 

2. We have the large hosted runners enabled. According to [GitHub docs](https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions#per-minute-rates) 1 minute of the ARM MacOS runner costs $0.16. When I was testing this in my own account I had to set my spending limit to a non-zero amount to make the ARM runner available.

### CODE REVIEW

File: .github/workflows/rebuild_macos_arm64.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+jobs:
+  rebuild:
+    runs-on: macos-13-xlarge
+    steps:
+      - uses: actions/checkout@v3
+      - name: Install tooling
+        run: |
+          brew install make cmake gcc
+      - name: Go to the build directory
```
Comment (puzpuzpuz) [inline .github/workflows/rebuild_macos_arm64.yml:15]: nit: the name is slightly misleading. Something like Build would be better

Comment (jerrinot) [reply .github/workflows/rebuild_macos_arm64.yml:15]: good eyes. that's a left-over from testing.

**Review by puzpuzpuz — COMMENTED** [review]
Looks like a nice improvement to me. Considering that ARM runners may cost something, let's hear from @bluestreak01

**Comment by bluestreak01** [comment]
we have M2 mac mini in the office, Marius just doesn't have time to set it up. Altho this is useful, this setup is different from what we're doing now. There is script to kick off cross-compile, which doesn't do ARM mac. Ideally this very same workflow should also build M1 binary.

Otherwise this is no different to fetching M1 laptop that building binary (more realibly than CI).

What do you guys think?

**Comment by bluestreak01** [comment]
@jerrinot can you build all binaries like this?

**Comment by jerrinot** [comment]
@bluestreak01 yeah, I see no reason why not: GitHub has a hosted MacOS Intel runner too and all other builds are cross-compiled on a Linux Intel machine. So it should be all doable. It would be nice to be able to compile all native libs with a single action, rather than messing with S3, different AWS accounts, Azure Personal Token, etc. So let me try it.

**Comment by jerrinot** [comment]
@bluestreak01 @puzpuzpuz: done. it's rebuilding native libs for all supported platforms.

File: .github/workflows/rebuild_native_libs.yml [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+      - name: Save FreeBSD binary to Cache
+        uses: actions/cache/save@v3
+        with:
+          path: |
+            core/src/main/resources/io/questdb/bin/freebsd/libquestdb.so
+          key: nativelibs-freebsd-${{ github.sha }}
+  build-macos:
+    strategy:
+      matrix:
+        # macos-12-xlarge = ARM M1
```
Comment (puzpuzpuz) [inline .github/workflows/rebuild_native_libs.yml:50]: nit:
```suggestion
        # macos-13-xlarge = ARM M1
```

File: .github/workflows/rebuild_native_libs.yml [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+          key: nativelibs-armosx-${{ github.sha }}
+      - name: Save x64 MacOS binary to Cache
+        if: ${{ matrix.os == 'macos-latest' }}
+        uses: actions/cache/save@v3
+        with:
+          path: |
+            core/src/main/resources/io/questdb/bin/osx/libquestdb.dylib
+          key: nativelibs-osx-${{ github.sha }}
+  build-others:
+    runs-on: ubuntu-latest
```
Comment (puzpuzpuz) [inline .github/workflows/rebuild_native_libs.yml:81]: Which glibc min version shall we have here? If I'm not mistaken, currently we use Ubuntu 20.04 to make sure that we're compatible with glibc 2.29.

Comment (jerrinot) [reply .github/workflows/rebuild_native_libs.yml:81]: good question! Azure Pipelines uses ubuntu-20.04, let me try the same here.

Comment (jerrinot) [reply .github/workflows/rebuild_native_libs.yml:81]: testing here: https://github.com/jerrinot/questdb/actions/runs/6875301619

Comment (jerrinot) [reply .github/workflows/rebuild_native_libs.yml]: the testing build passed 🎉

Comment (puzpuzpuz) [reply .github/workflows/rebuild_native_libs.yml]: nit: I'd add a comment on why we use this particular version, so that no one changes it accidentally.

**Review by puzpuzpuz — APPROVED** [review]
PR of the year!

**Review by bluestreak01 — CHANGES_REQUESTED** [review]
@jerrinot this is awesome!

But could you confirm GLIBC version we build linux binary with?

**Comment by jerrinot** [comment]
@bluestreak01 what version do we need? GitHub Actions [doc says](https://github.com/actions/runner-images/blob/ubuntu20/20231030.2/images/linux/Ubuntu2004-Readme.md) that Ubuntu 20.04 uses glibc 2.31. Which I confirmed in a [test run](https://github.com/jerrinot/questdb/actions/runs/6877663747/job/18705701696):
image

I have yet to test it, but I am asusming Azure Pipelines uses the same version - because it's [running Ubuntu 20.04](https://github.com/questdb/questdb/blob/bd8fcffccff4e8994045059df482666cac0cf7d3/ci/cbinary-pipeline.yml#L20) too. 
OTOH, @puzpuzpuz [says](https://github.com/questdb/questdb/pull/3953/#discussion_r1393864243) we should use 2.29 😕  So let me check what Azure actually has.

**Comment by jerrinot** [comment]
tl;dr: The Linux binaries produced by this GitHub Action are linking to the same glibc versions as the binaries produced by Azure Pipelines. In both cases, the version 2.17 should be sufficient.  

The long version:
What we have right now in `master`:
**x64**
```
$ strings ./linux/libquestdb.so|grep GLIBC|grep -v '@' 
GLIBC_2.2.5
GLIBC_2.9
GLIBC_2.12
GLIBC_2.3
GLIBC_2.14
GLIBC_2.3.2
GLIBC_2.4
GLIBC_2.17
GLIBC_2.7
GLIBC_2.3.4
```

**arm**
```
$ strings ./armlinux/libquestdb.so|grep GLIBC|grep -v '@'
GLIBC_2.17
```

The result of a [test build](https://github.com/jerrinot/questdb/actions/runs/6877663747/job/18705701696): 
**x64**
$ strings ./x64_libquestdb.so|grep GLIBC|grep -v '@'
```
GLIBC_2.2.5
GLIBC_2.9
GLIBC_2.12
GLIBC_2.3
GLIBC_2.14
GLIBC_2.3.2
GLIBC_2.4
GLIBC_2.17
GLIBC_2.7
GLIBC_2.3.4
```

**arm**
$ strings ./arm64_libquestdb.so|grep GLIBC|grep -v '@'
GLIBC_2.17

**Comment by bluestreak01** [comment]
cool, we need these binaries to work on AWS (2.26) and customer sites (2.29)

**Comment by bluestreak01** [comment]
can you actually try these version on AWS pls?

**Comment by bluestreak01** [comment]
we buld linux binaries with 2.31, it seems to be compatible

```
╭─vlad at DESKTOP-Q1FK2VL in ~
╰─○ ldd --version
ldd (Ubuntu GLIBC 2.31-0ubuntu9) 2.31
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
╭─vlad at DESKTOP-Q1FK2VL in ~
```

**Comment by jerrinot** [comment]
@bluestreak01 what do you mean by "testing on AWS"? Do you want me to test QuestDB with native binaries from this GitHub action on a particular Amazon Linux version?

**Comment by jerrinot** [comment]
@bluestreak01: tested on glibc 2.26 and 2.35 - works on both of them. I will add another CI job to test on various distributions, but I'd prefer to do it in a separate PR.
